### ISSUE TITLE
data passing and processing enhancements based on current data input/output design

### ISSUE THREAD

**Issue by hongchaodeng** [oam-dev/kubevela#770]
There have been some offline discussion about enhancing data input-output design.
In this issue,  I'm gonna summarize all discussions and converge everyone to a final solution here.

## Proposal

The existing data input/output feature supports **ordering and data passing** between resources (Components, Traits) within an AppConfig. However, in our production experience we have encountered more complex problems:

## Enhancement 1: Patching data value via JSON patch

The use case goes as below:

- The user reads a Redis config "instanceID: xyz"
- The user wants to fill a ConfigMap with key "redis" with above data:

  <details>
    <summary>ConfigMap</summary>

  ```yaml
  apiVersion: v1
  kind: ConfigMap
  data:
    redis: |
      {
        "instanceID":"xyz"
      }
  ```
  </details>

- Then the user reads another Redis config "instanceClass: abc"
- This time, the user wants to **patch** the ConfigMap with key "redis" with above data:
  <details>
    <summary>ConfigMap</summary>

    ```yaml
    apiVersion: v1
    kind: ConfigMap
    data:
      redis: |
        {
          "instanceID":"xyz"
        }
    data:
      redis: |
        {
          "instanceID":"xyz",
          "instanceClass": "abc"
        }
    ```
  </details>

The problem is that current data passing technique only supports single field. Therefore, the best it can pass is the "redis" field. Moreover, the supported types of the field's value are hardcoded in the code. Therefore, we couldn't do complex operations like JSON Patch.

To solve this problem, we would need to enhance data input-output with a new field called "operations". In "operations", we can add new mechanism to handle complex operations like JSON Patch. Below is an example:

<details>
 <summary>`operations` field</summary>

  ```yaml
  operations:
  - type: "jsonPatch"
    op: "add"
    fieldPath: "data"
    value: "{}"
  - type: "jsonPatch"
    op: "add"
    fieldPath: "data.redis"
    value: '"{}"'
  - type: "jsonPatch"
    op: "add"
    fieldPath: "data.redis"
    path: "instanceID"
    valueFrom: 
      fieldPath: "status.instanceID"
  - type: "jsonPatch"
    op: "add"
    fieldPath: "data.redis"
    path: "instanceClass"
    valueFrom: 
      fieldPath: "status.instanceClass"
  ```
</details>

## Enhancement 2: Output to and intput from intermediate data store

The use case goes as below:

- The user wants to create a Redis resource.
- The user prepares an empty ConfigMap.
- The user wants the Redis resource status fields to patch into the previous ConfigMap.
- The user wants to get the data from the ConfigMap once it is ready, and use it to set env of the application container.

Unlike tranditional data input-output scenarios where only two resources are involved, a new intermediate store has poped up in this new scenario. Normally users who creates backend services like MySQL/Redis would usually store the credentials into an intermediate store like Secret/ConfigMap, because they could reuse it for other purposes like UI, manual operations. Thus, we need to describe such intermediate store in data input-output workflow.

A critical question here is why not interpolate ConfigMap to a Trait or Component in the AppConfig? Based on current discussion and observation, the ConfigMap is not created or managed by the applicaiton owner. Thus it doesn't belong to the AppConfig.

A complete solution would be like below:

<details>
 <summary>AppConfig + Container example</summary>

  ```yaml
  apiVersion: core.oam.dev/v1alpha2
  kind: ApplicationConfiguration
  metadata:
    name: my-app-deployment
    annotations:
      version: v1.0.0
  spec:
    components:
      - componentName: my-test-component

        dataOutputs:
        - name: my-test-output

          store:
            apiVersion: v1
            kind: ConfigMap
            name: oam-redis-config
        
          conditions: 
          - op: "notEq"
            value: ""
            fieldPath: "status.instanceID"
          - op: "notEq"
            value: ""
            fieldPath: "status.instanceClass"

          operations:
          - type: "jsonPatch"
            op: "add"
            fieldPath: "data"
            value: "{}"
          - type: "jsonPatch"
            op: "add"
            fieldPath: "data.redis"
            value: '"{}"'
          - type: "jsonPatch"
            op: "add"
            fieldPath: "data.redis"
            path: "instanceID"
            valueFrom: 
              fieldPath: "status.instanceID"
          - type: "jsonPatch"
            op: "add"
            fieldPath: "data.redis"
            path: "instanceClass"
            valueFrom: 
              fieldPath: "status.instanceClass"

      - componentName: my-test-input-component
        dataInputs:
        - store:
            apiVersion: v1
            kind: ConfigMap
            name: oam-redis-config

          conditions:
          - op: "notEq"
            value: ""
            fieldPath: "data.redis"

          operations:
          - type: "jsonPatch"
            op: "add"
            fieldPath: "spec.containers[0].env[0].value"
            valueFrom:
              fieldPath: "data.redis"

  ---
  # The application container will receive the final data
  spec:
    containers:
    - env:
      - name: CACHE_SECRET
        value: '{"instanceClass":"...","instanceID":"..."}'
  ```
</details>

## (Optional) Data Centric API

The above proposal is based on existing API and not trying to introduce huge change. But ideally, if we have a data centric API like below, it would be much easier to model the intermediate store case. Below is an example:

<details>
 <summary>Data api</summary>

  ```java
  data: redis_conn: {
    from: {
      Component: "redis"
    }
    to: {
      component: "web"
    }
    saved: {
      secret: "redis-secret"
    }
  }
  ```
</details>

This is related to the CUE change of https://github.com/oam-dev/kubevela/issues/656 . Might be relevant to consider in that.

**Comment by wonderflow** [comment]
We're proposing to move all model layer feature to Definition files, what about let's move it to WorkloadDefinition and TraitDefinition?

```
kind: WorkloadDefinition
metadata:
  name: rds
spec:
    dataOutputs:
      - name: my-test-output
        store:
          apiVersion: v1
          kind: ConfigMap
          name: oam-redis-config
        conditions: 
        - op: "notEq"
          value: ""
          fieldPath: "status.instanceID"
        - op: "notEq"
          value: ""
          fieldPath: "status.instanceClass"
        operations:
        - type: "jsonPatch"
          op: "add"
          fieldPath: "data"
          value: "{}"
        - type: "jsonPatch"
          op: "add"
          fieldPath: "data.redis"
          value: '"{}"'
        - type: "jsonPatch"
          op: "add"
          fieldPath: "data.redis"
          path: "instanceID"
          valueFrom: 
            fieldPath: "status.instanceID"
        - type: "jsonPatch"
          op: "add"
          fieldPath: "data.redis"
          path: "instanceClass"
          valueFrom: 
            fieldPath: "status.instanceClass"
---
kind: WorkloadDefinition
metadata:
  name: web
spec:
      dataInputs:
      - store:
          apiVersion: v1
          kind: ConfigMap
          name: oam-redis-config

        conditions:
        - op: "notEq"
          value: ""
          fieldPath: "data.redis"

        operations:
        - type: "jsonPatch"
          op: "add"
          fieldPath: "spec.containers[0].env[0].value"
          valueFrom:
            fieldPath: "data.redis"

---
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: my-app-deployment
  annotations:
    version: v1.0.0
spec:
  components:
    - componentName: my-test-component
    - componentName: my-test-input-component


```

**Comment by hongchaodeng** [comment]
@wonderflow In your proposal, it is leaking the instance name of ConfigMap to Definition which is a schema:

```
        store:
          apiVersion: v1
          kind: ConfigMap
          name: oam-redis-config
```

**Comment by resouer** [comment]
The ConfigMap seems to be a medium resource? Then its name could be generated based on component name.

If the user indent to provide a ConfigMap instance, we have to leak its name anyway, either in schema or abstraction. It's basically not part of the model.

**Comment by resouer** [comment]
A off-topic thinking is, definition is not a schema essentially, it would reference schema/template, and I can image it may also reference other things (which could be configmap or secret instances) in the future.

**Comment by hongchaodeng** [comment]
> Then its name could be generated based on component name.

This is not possible. Please see top level description that "Based on current discussion and observation, the ConfigMap is not created or managed by the applicaiton owner. Thus it doesn't belong to the AppConfig."

> If the user indent to provide a ConfigMap instance, we have to leak its name anyway, either in schema or abstraction. It's basically not part of the model.

@resouer  I am not sure what you mean by this. Can you provide an example? 
The name of the store is something that breaks the schema nature of data input-output.
Not sure what we will do in the future, but moving data input into definition will have to decouple store and other parts at that point.

**Comment by resouer** [comment]
@hongchaodeng I see, so this ConfigMap is not a medium resource. 

In that case, since we can't model store such as ConfigMap and Secret in OAM for now, the name and kind of such resource will always be a "leak". Do we have plan to fix this?

**Comment by hongchaodeng** [comment]
The top level comment describes the current plan. This is how it is implemented now to support real use cases.

For the long term,  we will see how the final definition design will be and how data input-output with intermediate store is used at that time. Then we will design based on that.

If you want to discuss theoretical design I also have a "Data Centric API" which will better serve the template architecture and easier to model the intermediate design.

**Comment by resouer** [comment]
hongchaodeng The data centric api looks nice with cue template. Though I 'd say let's keep current design for now, it seems to be the best option at current stage.

**Comment by resouer** [comment]
For the long term, I'd like to have a clearer entrypoint for to define dataInput/Ouput. We need to discuss whether give a abstraction for end users to define it, or make it platform builder facing stuff and move it to definitions somehow. Or even need to do BOTH. /cc @wonderflow @hongchaodeng

**Comment by wonderflow** [comment]
Actually the data output can be easily defined in Definition:

```
kind: WorkloadDefinition
metadata:
  name: rds
spec:
    ...
    dataOutputs:      
      cueTemplate: |
         condition: {
            if context.status.instanceID == _|_ || context.status.instanceID == " "
                || context.status.instanceClass != _|_ || context.status.instanceClass != " " {
                  false
            }
         }
         output: {
            apiVersion: "v1"
            kind: "ConfigMap"
            spec: data:{
                redis: {
                  instanceID: context.status.instanceID
                  instanceClass: context.status.instanceClass
                }
            }
         }
```

1. KubeVela Runtime will get the real underlying workload CR as context.
2. KubeVela Runtime can check definition with the dataoutput.cueTemplate.condition, if the value is false, the condition was not meet the requirement, and we should periodically check.
3. When the condition meet requirement, output the template and create the resource with **a FIXED NAME such as `<componnetName>-data`**

**Comment by wonderflow** [comment]
For application level data passing, I have come up with a new proposal in #873

**Comment by wonderflow** [comment]
fixed by #1013
